openraft/src/engine/log_id_list.rs

```diff
@@ -0,0 +1,105 @@
+use crate::raft_types::RaftLogId;
+use crate::LogId;
+use crate::NodeId;
+
+/// Efficient storage for log ids.
+///
+/// It stores only the ids of log that have a new leader_id. And the `last_log_id` at the end.
+/// I.e., the oldest log id belonging to every leader.
+///
+/// If it is not empty, the first one is `last_purged_log_id` and the last one is `last_log_id`.
+/// The last one may have the same leader id as the second last one.
+#[derive(Default, Debug, Clone)]
+#[derive(PartialEq, Eq)]
+pub struct LogIdList<NID: NodeId> {
+    key_log_ids: Vec<LogId<NID>>,
+}
+
+impl<NID: NodeId> LogIdList<NID> {
+    pub fn new(key_log_ids: impl IntoIterator<Item = LogId<NID>>) -> Self {
+        Self {
+            key_log_ids: key_log_ids.into_iter().collect(),
+        }
+    }
+
+    /// Extends a list of `log_id` that are proposed by a same leader.
+    ///
+    /// The log ids in the input has to be continuous.
+    pub(crate) fn extend_from_same_leader<'a, LID: RaftLogId<NID> + 'a>(&mut self, new_ids: &[LID]) {
+        if let Some(first) = new_ids.first() {
```

Comment:
may be add some judge that new_ids term >= last log id term?

Reply:
Since it calls `self.append()` the assertions in `append()` would be enough?